### PR TITLE
py-{numpy,scipy}: update to latest version

### PR DESCRIPTION
**Pull request by reneeotten**
This PR updates `py-scipy` and `py-numpy` to their latest upstream versions (1.7.0 and 1.21.0, respectively). 
For the SciPy update a few new dependencies are required as `pythran` is a build dependency; secondly it now uses `fetch.type git` and calls `git submodule` to get the vendored `boost` dependency. 

Please note that NumPy upstream states ```Warning: there are unresolved problems compiling NumPy 1.21.0 with gcc-11.1 .``` However, it seems to build fine and there are only a few test failures - 5 compared to 2 failures in v1.20.3; the four new failures are all in `typing/tests/test_typing.py`.

SciPy has some more issues in that the test-suite segfaults (in `optimize/tests/test_linprog.py`), but that also happens for both versions. It would be nice to fix that though, as it does work fine when installing with `pip` in a `virtualenv` or when using `conda`.

###### Tested on
macOS 10.15.7 19H1217 x86_64
Xcode 12.4 12D4e

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [X] tried existing tests with `sudo port test`? A few test-failures, see above
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

File: python/py-numpy/Portfile
Comment (michaelld) [inline python/py-numpy/Portfile:99]: I believe that this dependency is required for older NumPy used for older Python ... we need to test this out & if so then add it to the dependency list for those.

File: python/py-scipy/Portfile
Comment (michaelld) [inline python/py-scipy/Portfile:100]: Same per the first comment

File: python/py-scipy/Portfile
Comment (michaelld) [inline python/py-scipy/Portfile:106]: Same per the first comment

**Review by michaelld — CHANGES_REQUESTED** [review]
Thanks for the PR! I was updating these ports Friday & had no issues building NumPy (didn't make it to SciPy) them with no changes for dependency/ies ... but "test" hung ... and then I had a busy weekend .. and here we are!

Are few items to check / verify before we merge. Also: I've restarted "check" ... 2/3 failed, which isn't a good thing in general but I'm not sure if that's expected inside GH at the moment. I have not checked the MP CI yet.

File: python/py-numpy/Portfile
Comment (reneeotten) [reply python/py-numpy/Portfile:99]: right, I didn't check that sorry... but on the other hand I find it hard to believe that `nose` would actually be anything other than a test dependency. Since there is no test-phase defined in the Portfile it shouldn't be needed. 

Of course, if the intention here is to allow people to do `import numpy ; numpy.test()` then it could be needed. In that case it should make sure to have all test dependencies added (I had a missing `hypothesis` either here or in scipy). But that would benefit from a comment in the Portfile or we could just leave this up to the user...

File: python/py-scipy/Portfile
Comment (reneeotten) [reply python/py-scipy/Portfile:106]: that is a library dependency of `pytest`, so there isn't a reason I can think of to list it separately here

File: python/py-numpy/Portfile
Comment (michaelld) [reply python/py-numpy/Portfile:99]: good points! I guess I need a "test" phase then ... since I do that after updating and before pushing the updates ... just to make sure testing does something useful.

File: python/py-scipy/Portfile
Comment (michaelld) [inline python/py-scipy/Portfile:87]: From the main info here: `it now uses fetch.type git and calls git submodule to get the vendored boost dependency.` ... let me do some testing & see what's what here. I really don't like having to go this route unless it's absolutely required to do so.

**Comment by michaelld** [comment]
I'm OK with all of the changes except those to SciPy, which is just that I need more time to do testing / verification / see if there's a better way for the Boost dependency

File: python/py-scipy/Portfile
Comment (reneeotten) [reply python/py-scipy/Portfile:87]: Sure - feel free! I am not a big fan either and would prefer to use the MP version of boost. But I couldn't find a very easy way of specifying it and admittedly didn't attempt to start patching stuff. Also, it might be wise to stick with the vendored version as the SciPy folks have tested this.

Comment (michaelld) [reply python/py-scipy/Portfile:87]: Yes sticking with a specific Boost is often the best thing to do ... hopefully the new Boost MP options will be good enough for what SciPy needs

Comment (reneeotten) [reply python/py-scipy/Portfile:87]: the latest commit in the SciPy repository regarding this is :`REV: revert to Boost 1.75 to avoid constexpr bug with gcc <7.1 `; so I guess we don't have a `boost` version in MP that is compatible with it. 

I did consider switching to PyPI as `master_sites` since the files there presumably contain the vendored `boost` library. One would probably need to switch then to PyPI for the older versions of SciPy as well. I think that should work, but haven't tested it.

Comment (michaelld) [reply python/py-scipy/Portfile:87]: Using MP boost will not be difficult. For this release, the submodule is 1.76.0 ... regardless of the GCC issue as you note. It's straight forward to allow MP to select the Boost version, so moving back to 1.75.0 or 1.71.0 or whatever is doable ... if it works with this SciPy of course :)

Comment (reneeotten) [reply python/py-scipy/Portfile:87]: I agree it's easy to allow MP to select the Boost versions, the question is how easy it is to tell SciPy to actually use that ;)

Comment (michaelld) [reply python/py-scipy/Portfile:87]: Pretty easy it turns out ...

Comment (michaelld) [reply python/py-scipy/Portfile:87]: works with Boost 1.71.0 ... so we're good to go without having to create more Boost ports

**Comment by michaelld** [comment]
[py-scipy_boost_STDC.zip](https://github.com/macports/macports-ports/files/6736437/py-scipy_boost_STDC.zip)

**Comment by michaelld** [comment]
@reneeotten Can you try the patches I just attached? Those should fix the Boost usage so-as to avoid the GIT stuff. It also fixes the <threads.h> issue by disabling STDC THREADS and using __thread instead. It's not clear to me why this is necessary, but the patch fixes the issue for me.

**Comment by michaelld** [comment]
If those patches work, then I'd say to go ahead and merge them into this PR & force push over this with the new port files

**Comment by reneeotten** [comment]
thanks @michaelld  - I'll take a look and will update this PR once I've had the chance. Would you like me to add a test-phase to both ports at the same time as well?

**Comment by michaelld** [comment]
Yes please! In my Portfile I did shuffle the testing dependencies to be together, but I didn't take the next step of actually adding in a testing phase. Thx!

**Comment by reneeotten** [comment]
> Yes please! In my Portfile I did shuffle the testing dependencies to be together, but I didn't take the next step of actually adding in a testing phase. Thx!

the Portfile you added in the attachment is actually for `cmake` - so that's not all that helpful ;)

**Comment by michaelld** [comment]
LOL oops! I'll drop the correct one later at work. Sorry about that!

**Comment by michaelld** [comment]
[Portfile.txt](https://github.com/macports/macports-ports/files/6741036/Portfile.txt)

**Comment by michaelld** [comment]
Double checked that one before attaching it :)

**Comment by reneeotten** [comment]
@michaelld your patches do work indeed. I am not sure what the purpose of the "threads" patch is or whether that is needed, but if you say so I'll take your word for it. 

I removed the test dependencies and decided against adding a test-phase. I don't think this is actually beneficial since you would need to build the port twice (once for the testing and then for installing it), which in my opinion is a waste of time. So I think one should just install the port and then run the tests from the Python interpreter (with `import scipy ; scipy.test('full')` or the equivalent for numpy). In that case adding dependencies as `depends_test` does not help since they will be only used in the test-phase, but adding them as `depends_run` or `depends_lib` is incorrect as they are not really needed. So let's just leave this up to the user to install the dependencies needed for testing.

**Review by michaelld — APPROVED** [review]
LGTM ; thx!

**Comment by michaelld** [comment]
I can't figure out why the CI / checks are failing, but all of the changes here look solid now, so I'm going ahead with the merge.

**Pull request merged**
